### ISSUE TITLE
DeepSeek V4 ignores packed-document boundaries in sparse attention

### ISSUE THREAD

**Issue by HaoranYu111** [pytorch/torchtitan#4801]
When packing independent training documents with position IDs reset at each document, DeepSeek V4 accepts those position IDs, but its sparse attention window can cross a document boundary. With two 64-token documents, query 64 (position 0 of the second document) is allowed to attend to key 63 (the last token of the first document).

This concerns packed samples that are intended to remain isolated. Single-document input and training that intentionally allows cross-document attention are outside this report.

The shared decoder treats position resets as document boundaries. DSV4 [overrides `get_attention_masks()`](https://github.com/pytorch/torchtitan/blob/6c2dadbb3109998092d1164e990c6377b2b8fb50/torchtitan/models/deepseek_v4/model.py#L222-L232) and discards positions, padding and document metadata. Its sparse window is then constructed from offsets in the entire flattened token stream.

Reproduced on TorchTitan [`6c2dadb`](https://github.com/pytorch/torchtitan/commit/6c2dadbb3109998092d1164e990c6377b2b8fb50), with PyTorch 2.12.0+cpu and spmd_types 0.2.5. The CPU reproduction below calls the current model mask hook and constructs the actual DSV4 block mask. It reports:

```text
positions = torch.arange(64).repeat(2)
query = 64, key = 63
DSV4 sparse mask: True
shared packed-document mask: False
single-document control: True
```

This checks mask semantics without running FlexAttention kernels or a full training job. I have not tested the latest PyTorch nightly or quantified a training-loss impact.

Could we either support packed-document boundaries throughout the DSV4 attention/compressor path, or reject unsupported packed input explicitly? Fixing only the sliding window would not establish correct compressed-KV and indexer behavior. A regression should also check that changing one document leaves another document's output unchanged.

Related: [#4452](https://github.com/pytorch/torchtitan/pull/4452) changes the CSA implementation. At `e203ecd`, its mask hook uses the shape and device of `positions`, but does not inspect position resets. It may be a useful place to coordinate support.

<details>
<summary>CPU reproduction (run from the pinned TorchTitan checkout)</summary>

```python
import json

import torch

from torchtitan.models.common.attention import (
    get_efficient_causal_mask_mod_for_packed_document,
)
from torchtitan.models.deepseek_v4.attention import DSV4FlexInnerAttention
from torchtitan.models.deepseek_v4.model import DeepSeekV4Model

# Two packed documents, each with 64 tokens.
positions = torch.arange(64).repeat(2)
assert DeepSeekV4Model.get_attention_masks(None, positions) is None
inner = DSV4FlexInnerAttention.Config(
    block_size=128,
    window_size=16,
    compress_ratio=1,
    softmax_scale=0.1,
    index_topk=0,
).build()
selected = inner.get_window_topk_idxs(bsz=1, seqlen=128, device="cpu")
selected = torch.cat(
    (selected, torch.full((1, 128, 1), 128, dtype=torch.int64)), dim=-1
)
mask = inner._build_block_mask(1, 128, 129, selected, "cpu")
idx = [torch.tensor(v) for v in (0, 0, 64, 63)]
actual = mask.mask_mod(*idx).item()
expected = get_efficient_causal_mask_mod_for_packed_document(positions)(*idx).item()
assert actual is True and expected is False
# A single-document input keeps the same local causal window.
single_expected = get_efficient_causal_mask_mod_for_packed_document(torch.arange(128))(
    *idx
).item()
assert single_expected is True
print(
    json.dumps(
        {
            "packed_positions": "arange(64).repeat(2)",
            "query_token": 64,
            "key_token": 63,
            "dsv4_allows_cross_document": actual,
            "common_document_mask_allows": expected,
            "single_document_control_allows": single_expected,
            "scope": "current model mask hook and actual sparse mask; no attention kernel or model training",
        }
    )
)
```

</details>

**Comment by tianyu-l** [comment]
cc @drisspg @floatingtrees
